sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc

```diff
@@ -0,0 +1,506 @@
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Load API
+
+==== Single Value Overload
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+// Load scalar
+template<typename Group, typename InputIteratorT,
+         typename OutputT, typename Properties = empty_properties_t>
+void group_load(Group g, InputIteratorT in_iter,
+                      OutputT& out, Properties = {});
+}
+----
+
+_Preconditions_: `in_iter` must be the same for all work-items
+in the group.
+
+._Constraints_:
+* Only available if `Group` is a work-group or sub-group.
+* `InputIteratorT` suppose to be a random access iterator.
+* Value type of `InputIteratorT` must be convertible to `OutputT`.
+
+_Effects_: Loads single element from `in_iter` to `out` by using the `g` group
+object to identify memory location as `in_iter` + `g.get_local_linear_id()`.
+
+`Properties` argument is reserved for future revisions of this extention and is
+ignored now. Default value is empty
+`sycl::ext::oneapi::experimental::empty_properties_t`.
+May be used in future for setting boundary values or limiting numbers of work
+items.
+
+==== `sycl::vec` Overload
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+// Load API sycl::vec overload
+template<typename Group, typename InputIteratorT,
+         typename OutputT, int N, typename Properties = empty_properties_t>
+void group_load(Group g, InputIteratorT in_iter,
+                      sycl::vec<OutputT, N>& out, Properties = {});
+}
+----
+
+_Preconditions_: `in_iter` must be the same for all work-items
+in the group.
+
+._Constraints_:
+* Only available if `Group` is a work-group or sub-group.
+* `InputIteratorT` suppose to be a random access iterator.
+* value type of `InputIteratorT` and `OutputT` types must be convertible basic
+scalar types supported in device code.
+
+_Effects_: Loads `N` elements from `in_iter` to `out`
+using the `g` group object.
+Properties may specify xref:data_placement[data placement].
+Default data placement is a blocked one:
+`out[i]` = `in_iter[g.get_local_linear_id() * N + i];`
+in striped case:
+`out[i]` = `in_iter[g.get_local_linear_id() + g.get_local_linear_range() * i];`
+for `i` between `0` and `N`.
+
+==== Fixed-size Array Overload
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+// Load API
+template<typename Group, typename InputIteratorT, typename OutputT,
+        std::size_t ElementsPerWorkItem, typename Properties = empty_properties_t>
+void group_load(Group g, InputIteratorT in_iter,
+                sycl::span<OutputT, ElementsPerWorkItem> out, Properties = {});
+}
+----
+
+._Preconditions_:
+* `in_iter` must be the same for all work-items in the group.
+* `out` must be a `sycl::span` made from a pointer to the private memory space.
```

Comment:
This is a good catch. I think we should probably clarify that the other interfaces are required to be allocated in the private address space too.

The original thinking here was that these functions are intended to be used to load from/store to memory.  If the arguments could be in the local or global memory space, the functionality would be closer to a "copy" operation.

Reply:
Is there really any reason to limit it to the private address space, though?  If the implementation works for any address space, why add a limitation?

With the existing API, we could do something like this:

```
acc[i] = sg.load(ptr);
```

The most natural way to write this in the new API would be:

```
group_load(sg, ptr, acc[i]);
```

However, that would violate the precondition because `acc[i]` is not in the private address space.  Instead, a conformant application would need to write:

```
T x;
group_load(sg, ptr, x);
acc[i] = x;
```

Wouldn't it be easier to just allow the destination of the load to be in any address space?

Reply:
I don't have a strong opinion here from API point of view, but I assume for performance value it makes sense to guarantee that it is private memory, as I am not sure, that we can check it inside of the implementation.
@aelovikov-intel, could you please, advise here?

Reply:
The implementation could be written to ensure that the SPIR-V instruction does always access private memory.  For example (pseudo-code):

```
void group_load(Group g, InputIteratorT in_iter, OutputT& out) {
  OutputT tmp = __spirv_SubgroupBlockReadINTEL(in_iter);
  out = tmp;
}
```

I assume the optimizer will eliminate the use of `tmp`.

Reply:
> ```
> acc[i] = sg.load(ptr);
> ```

This is doing two things:

1. `sg.load(ptr)` loads from `ptr` into a variable (in private memory) and returns it.
2. `=` assigns the value from private memory to `acc[i]`

> ```
> T x;
> group_load(sg, ptr, x);
> acc[i] = x;
> ```

This is definitely more verbose, but I like that it clearly remains as two steps.  Somebody might otherwise expect `group_load(sg, ptr, &acc[i])` to issue something like a DMA to "load" directly into the memory associated with `acc[i]`.


> Wouldn't it be easier to just allow the destination of the load to be in any address space?

If we did this, I think we'd want to say in the specification that if the pointer is in the local or global address space, the implementation will/can load into private memory first.

But I'm still not sure this is a good idea.  Consider what this looks like for the store version of things:

```c++
group_store(g, &acc[i], ptr);
```

...would people expect that to load the value from `acc[i]` into private memory, then write it out to `ptr`?  Compare with:

```c++
T x = acc[i];
group_store(g, &x, ptr);
```

Reply:
> The implementation could be written to ensure that the SPIR-V instruction does always access private memory. For example (pseudo-code)

It has to be written like that to satisfy the implicit conversion to the output type that the extension mandates (or at least used to in previous revisions).

Reply:
> to satisfy the implicit conversion to the output type that the extension mandates (or at least used to in previous revisions)

The language about conversion is still in the spec.

Reply:
The reason I worry about adding the precondition about the private address space is that I think it's easy to miss.  The compiler won't diagnose an error, and I think our implementation will run correctly even if you pass a reference to a different address space.  I think this will lead to a lot of code that is non-conforming.  If our implementation ever changes to actually require the private address space, this code will suddenly break.

If we really want to require the private address space, I think we should change the parameter to be a `multi_ptr` with `address_space::private_space`.  That will force people to think about the address space, and the compiler will enforce it.  However, this also means that the common code pattern where you load into a block-scope variable will become much more verbose.

Given @aelovikov-intel's comment, it seems like we need the internal temporary variable anyway, so it seems like the implementation will end up allowing the parameter to be a reference to any address space anyways.

Reply:
I think the behavior you described is exactly why this ended up as a precondition.  It's UB, so anything can happen (including that it works).  We can't prove that it's a private pointer in the general case (e.g., for generic pointers).  But I agree that this is going to result in lots of bugs that are hard to track down.

I really don't want developers to have to use a `multi_ptr` to get good performance here, so you've convinced me that removing the restriction is a better solution.

How do you feel about my suggestion that we document that implementations can/will stage things via private memory?  I think we're going eventually going to end up with a `group_copy` function too, and I want to make sure that developers understand the difference.

Reply:
ok, I removed private mentioning from preconditions, thank you!

Reply:
> How do you feel about my suggestion that we document that implementations can/will stage things via private memory?

This seems like an implementation detail to me.

> I think we're eventually going to end up with a `group_copy` function too, and I want to make sure that developers understand the difference.

How would `group_copy` functions be different?  In particular, the "Fixed-size Array Overload" from this spec seems like it is copying memory from an iterator to another container.

Reply:
> How would `group_copy` functions be different? In particular, the "Fixed-size Array Overload" from this spec seems like it is copying memory from an iterator to another container.

I think the difference is mainly intent: the developer can be clearer about what they're trying to express, and the implementation can have different behavior that's aligned with the developer's intent.

`group_load` was intended to mean "load this value from memory, in the most efficient way possible".  Developers should expect it to act as a (potentially faster) drop-in replacement for something like running `T x = *ptr;`, on each work-item.  The reason the function exists is to help implementations reason about cases where all the work-items in the group are accessing `ptr` in a pattern that can be optimized.  Conceptually, the developer describes how all the accesses made by individual work-items can be combined into a larger block.  Each work-item owns a dedicated piece of the output, but collaborating with other work-items improves performance.

`group_copy` will mean "copy these values from memory X to memory Y, in the most efficient way possible".  Developers should expect it to act as a (potentially faster) drop-in replacement for something like `memcpy` on a single work-item.  Conceptually, the developer describes an operation that needs to happen (i.e., the copy), and the implementation gets to decide how that operation is distributed across work-items or whether it can employ dedicated hardware (e.g. DMAs).  Work-items don't own dedicated pieces of the output, but collaborate to move all the bytes from X to Y.

Reply:
Yes, I see the difference now.  With the load / store functions, each work-item provides its contribution to the values that are loaded or stored via a non-uniform parameter.  With the copy function, there would just be two uniform iterators, and it would be undefined which work-items copied which elements.